### PR TITLE
fix(eighth): fix duplicate signups when transferring students

### LINKED ISSUE #1051: Moving an activity causes duplicate signups
## Steps to reproduce
1. Two activities on two separate blocks
2. Have a student sign up for the two activities
3. Move one of the activities to the other block

## Expected behavior
Either one of the signups is dropped, or something is shown so the admin can choose which signup is kept.

## Actual behavior
A duplicate signup is created.

### PR DESCRIPTION
**Pull request by Laur04**
closes #1051

### CODE REVIEW

File: intranet/apps/eighth/views/admin/scheduling.py [hunk 1]
```diff
@@ -404,22 +406,44 @@ def transfer_students_action(request):
...
         else:
-            source_act.eighthsignup_set.update(scheduled_activity=dest_act)
+            duplicate_sign_up_users = []
+            duplicate_signups = []
+            with transaction.atomic():
+                for u in source_act.members.all():
+                    conflict_signup = u.eighthsignup_set.filter(scheduled_activity__block=dest_act.block)
+                    if conflict_signup.exists():
+                        duplicate_sign_up_users.append(u)
+                        duplicate_signups.append(conflict_signup[0])
```
Comment (theo-o) [inline intranet/apps/eighth/views/admin/scheduling.py:431]: While we hope that there are no duplicate signup bugs elsewhere in Ion, assuming that fact might not be a great idea.

**Review by theo-o — COMMENTED** [review]
This change may be worthy of unit tests to prevent future regressions and ensure consistent behavior

File: intranet/apps/eighth/tasks.py [hunk 2]
```diff
@@ -66,6 +66,49 @@ def room_changed_single_email(
     )
 
 
+@shared_task
+def transferred_activity_email(
+    dest_act: EighthScheduledActivity, duplicate_students: Collection[get_user_model()]  # pylint: disable=unsubscriptable-object
```
Comment (anonymoose2) [inline intranet/apps/eighth/tasks.py:71]:
```suggestion
    dest_act: EighthScheduledActivity, duplicate_students  # pylint: disable=unsubscriptable-object
```
Type annotations generally aren't supposed to contain function calls.

In the past, I've just omitted type annotations on parameters that reference `User` in `eighth/models.py` 
-- `eighth/models.py` depends on `users/models.py`, so importing it would create circular import issues. I've since learned that it's possible to work around the circular import issue, but that would probably be a matter for another PR.

File: intranet/apps/eighth/views/admin/scheduling.py [hunk 3]
```diff
@@ -404,22 +406,44 @@ def transfer_students_action(request):
...
+                source_act.eighthsignup_set.update(scheduled_activity=dest_act)
+
+            transferred_activity_email.delay(dest_act, duplicate_sign_up_users)
+
+            context["moved_students"] = duplicate_signups
 
             invalidate_obj(source_act)
             invalidate_obj(dest_act)
             messages.success(request, "Successfully transfered {} students.".format(num))
-        return redirect("eighth_admin_dashboard")
```
Comment (anonymoose2) [inline intranet/apps/eighth/views/admin/scheduling.py:420]: I would keep the redirect if there are no duplicates, but render a different "transfer_students_results" page if there are any duplicates.

File: intranet/apps/eighth/views/admin/scheduling.py [hunk 4]
```diff
@@ -404,22 +406,44 @@ def transfer_students_action(request):
...
-            source_act.eighthsignup_set.update(scheduled_activity=dest_act)
+            duplicate_sign_up_users = []
+            duplicate_signups = []
+            with transaction.atomic():
+                for u in source_act.members.all():
+                    conflict_signup = u.eighthsignup_set.filter(scheduled_activity__block=dest_act.block)
+                    if conflict_signup.exists():
+                        duplicate_sign_up_users.append(u)
+                        duplicate_signups.append(conflict_signup[0])
+                lock_on(duplicate_sign_up_users)
```
Comment (anonymoose2) [inline intranet/apps/eighth/views/admin/scheduling.py:432]: I would lock on **all** of the `User`s that are being transferred, to ensure that nothing happens concurrently.

**Review by anonymoose2 — CHANGES_REQUESTED** [review]

File: intranet/apps/eighth/views/admin/scheduling.py (hunk 1 above)
Comment (anonymoose2) [reply intranet/apps/eighth/views/admin/scheduling.py:431]: @theo-o what are you suggesting? Add all the users, then check for duplicates retroactively?

Comment (theo-o) [reply intranet/apps/eighth/views/admin/scheduling.py:431]: I am suggesting adding all elements of the `duplicate_signups` QuerySet, instead of just the first element.

File: intranet/apps/eighth/views/admin/scheduling.py (hunk 3 above)
Comment (Laur04) [reply intranet/apps/eighth/views/admin/scheduling.py:420]: I think it is useful to have the transfer messages appear beneath the original text and this stays in theme with many of the 8th admin functions, which all have the appearance of occurring on the same page. Redirecting to another page would also involve repeating a lot of code that is already in this view. I'd rather just redirect to "eighth_admin_dashboard" if there are no duplicates.

Comment (etnguyen03) [reply intranet/apps/eighth/views/admin/scheduling.py:420]: My concern is that such a "return the same-page" may cause an admin to press the "Transfer" button again; although that won't be a concern (as pressing the "Transfer" button again doesn't appear to break anything), but I think that it'd be better off to redirect to somewhere without a "Transfer" button.

Comment (Laur04) [reply intranet/apps/eighth/views/admin/scheduling.py:420]: Okay the latest rendition (just pushed) keeps it on the same page as far as django is concerned, but hides the "Transfer" section with template tags. This mimics the behavior of the rest of the page, which uses template tags to show/hide the wizard.

Comment (anonymoose2) [reply intranet/apps/eighth/views/admin/scheduling.py:420]: I'm happy with that. @etnguyen03 ?

File: intranet/templates/eighth/emails/transferred_activity.html [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<html><body>
+    {% with description="Room changes" target=info_link name="View status" %}
+        {% include "email_metadata.html" %}
+    {% endwith %}
+
+    <p>
+        You were transferred by the 8th Period Office into {{ dest_act.activity }} on {{ date_str }} during {{ dest_act.block.block_letter }} block. Your previous signup has been removed, but you may re-signup on Ion if you wish.
```
Comment (anonymoose2) [inline intranet/templates/eighth/emails/transferred_activity.html:7]:
```suggestion
        You were transferred by the 8th Period Office into {{ dest_act.activity }} on {{ date_str }} during {{ dest_act.block.block_letter }} block. Your previous signup during this block has been removed, but you may re-signup on Ion if you wish.
```
Just a minor clarification.

File: intranet/templates/eighth/emails/transferred_activity.txt [hunk 6]
```diff
@@ -0,0 +1,5 @@
+You were transferred by the 8th Period Office into {{ dest_act.activity }} on {{ date_str }} during {{ dest_act.block.block_letter }} block. Your previous signup has been removed, but you may re-signup on Ion if you wish.
```
Comment (anonymoose2) [inline intranet/templates/eighth/emails/transferred_activity.txt:1]:
```suggestion
You were transferred by the 8th Period Office into {{ dest_act.activity }} on {{ date_str }} during {{ dest_act.block.block_letter }} block. Your previous signup during this block has been removed, but you may re-signup on Ion if you wish.
```

File: intranet/apps/eighth/views/admin/scheduling.py (hunk 3 above)
Comment (etnguyen03) [reply intranet/apps/eighth/views/admin/scheduling.py:420]: Good to me as well.

File: intranet/templates/eighth/admin/transfer_students.html [hunk 7]
```diff
@@ -38,9 +48,14 @@
         {% else %}
             <p>Do you want to transfer <b>{{ num }} students</b> from:</p>
             <p><b>{{ source_act }}</b></p>
-            to:<br><br>
+            to:<br>
```
Comment (anonymoose2) [inline intranet/templates/eighth/admin/transfer_students.html:51]: Why was this `<br>` removed?

File: intranet/templates/eighth/admin/transfer_students.html [hunk 8]
```diff
@@ -30,6 +30,16 @@
...
             {% endcomment %}
         </form>
+    {% elif moved_students %}
+        <p>
+            The following students were already signed up for activities on {{ dest_act.block }}.
+            These signups have been replaced with {{ dest_act }}, and these students have been notified via email:
+        </p>
+        <ul>
+            {% for s in moved_students %}
+            <li>{{ s }}</li>
```
Comment (anonymoose2) [inline intranet/templates/eighth/admin/transfer_students.html:40]: I would want to print out the activity that each student was in. That way, if the admin who performs this action realizes this change messes something up, they can easily go through and revert the changes.

**Comment by etnguyen03** [comment]
One thing that I found: if you have two activities (call them X and Y) on blocks A and B (respectively), then if a user is signed up for X for A block and Y for B block, and someone starts a "transfer" from Y B block to X A block, then users will be emailed as if their A block signup changed, but when in reality only their B block signup changed.

So we may want to change the email to read something like "your B block (or the block from which you were transferred from) signup was cancelled; sign up before the deadline", or maybe we should just email everyone affected when a transfer happens?

File: intranet/templates/eighth/admin/transfer_students.html (hunk 8 above)
Comment (Laur04) [reply intranet/templates/eighth/admin/transfer_students.html:40]: This does. It prints the `EighthSignup` which includes the username, activity and block.

File: intranet/templates/eighth/admin/transfer_students.html (hunk 7 above)
Comment (Laur04) [reply intranet/templates/eighth/admin/transfer_students.html:51]: The space was weird. You can see it without a dev env at https://ion.tjhsst.edu/eighth/admin/scheduling/transfer_students_action?source_act=135085&dest_act=133549. This just moves that bottom line up.

**Comment by Laur04** [comment]
> One thing that I found: if you have two activities (call them X and Y) on blocks A and B (respectively), then if a user is signed up for X for A block and Y for B block, and someone starts a "transfer" from Y B block to X A block, then users will be emailed as if their A block signup changed, but when in reality only their B block signup changed.
> 
> So we may want to change the email to read something like "your B block (or the block from which you were transferred from) signup was cancelled; sign up before the deadline", or maybe we should just email everyone affected when a transfer happens?

We could make it even more general..."The eighth period office transferred you into a new activity on {{ block }}. This resulted in you being moved out of one or more activities that you previously signed up for. Please check your signups on Ion to ensure that they are what you want."

Since we are overriding email preferences here in order to give what we consider important information, I don't think we should email anyone except the students who had signups deleted.

**Comment by etnguyen03** [comment]
> > One thing that I found: if you have two activities (call them X and Y) on blocks A and B (respectively), then if a user is signed up for X for A block and Y for B block, and someone starts a "transfer" from Y B block to X A block, then users will be emailed as if their A block signup changed, but when in reality only their B block signup changed.
> > So we may want to change the email to read something like "your B block (or the block from which you were transferred from) signup was cancelled; sign up before the deadline", or maybe we should just email everyone affected when a transfer happens?
> 
> We could make it even more general..."The eighth period office transferred you into a new activity on {{ block }}. This resulted in you being moved out of one or more activities that you previously signed up for. Please check your signups on Ion to ensure that they are what you want."
> 
> Since we are overriding email preferences here in order to give what we consider important information, I don't think we should email anyone except the students who had signups deleted.

Well if 8PO (or someone) does it after someone signs up for eighth period, they may not check Ion until after the signup deadline and forget to sign up for an activity for which they were assigned out of.

I get the override email preferences but I still think that if a student is un-signedup from an activity, they should get an email. That being said, I'm not sure if 8PO will ever use the "transfer" to transfer the entire school out of a sticky activity (like, for instance, if they want to move the A block tomorrow to B block); if they do, then we'd be emailing the entire school which probably isn't a good idea to do. Maybe have a checkbox ("do you want to email everyone affected?") before sending - perhaps on the "confirm" page?

File: intranet/templates/eighth/admin/transfer_students.html (hunk 8 above)
Comment (anonymoose2) [reply intranet/templates/eighth/admin/transfer_students.html:40]: Ah. Missed that.

**Comment by Laur04** [comment]
Okay, how about this? To avoid the issue you brought up with sticky, school-wide activities while still ensuring that people don't get stuck in something they didn't want after the deadline, I will send an email to anyone who had an `EighthSignup` deleted in the transfer process, unless the activity associated with that `EighthSignup` is sticky. I'll reword the email per @anonymoose2 and your suggestions to be more clear about exactly what is going on.

**Comment by etnguyen03** [comment]
Maybe? Although I still believe that a checkbox is the better way to go (what about if it's a "sticky" counselor meeting was rescheduled - I remember that there are/were some mandatory counselor meetings that were marked as "sticky" throughout the year?)

File: intranet/templates/eighth/emails/transferred_activity.html [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<html><body>
+    {% with description="Room changes" target=info_link name="View status" %}
+        {% include "email_metadata.html" %}
+    {% endwith %}
+
+    <p>
+        You were transferred by the 8th Period Office into {{ dest_act.activity }} on {{ date_str }} during {{ dest_act.block.block_letter }} block. Your previous signup during this block and your signup during {{ source_act.block }} have been removed. Please visit Ion to ensure you are in the correct activities.
```
Comment (etnguyen03) [inline intranet/templates/eighth/emails/transferred_activity.html:7]:
```suggestion
        You were transferred by the 8th Period Office into {{ dest_act.activity }} on {{ date_str }} during {{ dest_act.block.block_letter }} block. Your previous signup during this block and your signup during {{ source_act.block }} have been removed. Please visit Ion to ensure you are signed up for the activities you are planning to attend.
```

File: intranet/templates/eighth/emails/transferred_activity.txt [hunk 10]
```diff
@@ -0,0 +1,5 @@
+You were transferred by the 8th Period Office into {{ dest_act.activity }} on {{ date_str }} during {{ dest_act.block.block_letter }} block. Your previous signup during this block and your signup during {{ source_act.block }} have been removed. Please visit Ion to ensure you are in the correct activities.
```
Comment (etnguyen03) [inline intranet/templates/eighth/emails/transferred_activity.txt:1]:
```suggestion
You were transferred by the 8th Period Office into {{ dest_act.activity }} on {{ date_str }} during {{ dest_act.block.block_letter }} block. Your previous signup during this block and your signup during {{ source_act.block }} have been removed. Please visit Ion to ensure you are signed up for the activities you are planning to attend.
```

**Review by etnguyen03 — CHANGES_REQUESTED** [review]
The logic looks fine to me; just a change to the email

File: intranet/apps/eighth/views/admin/scheduling.py [hunk 11]
```diff
@@ -402,24 +405,54 @@ def transfer_students_action(request):
     else:
         raise Http404
 
+    if "send_emails" in request.POST:
+        send_emails = True
```
Comment (anonymoose2) [inline intranet/apps/eighth/views/admin/scheduling.py:409]:
```suggestion
    send_emails = bool(request.POST.get("send_emails"))
```

File: intranet/templates/eighth/admin/transfer_students.html [hunk 12]
```diff
@@ -30,6 +30,16 @@
                 {% endif %}
             {% endcomment %}
         </form>
+    {% elif moved_students %}
+        <p>
+            The following students were already signed up for activities on {{ dest_act.block }}.
+            These signups have been replaced with {{ dest_act }}:
```
Comment (anonymoose2) [inline intranet/templates/eighth/admin/transfer_students.html:36]: It might be nice to reiterate whether the students have been notified via email, just for reference. i.e. say either "The students have been notified" or "The students have **not** been notified."

File: intranet/apps/eighth/views/admin/scheduling.py [hunk 13]
```diff
@@ -392,6 +394,7 @@ def transfer_students_action(request):
 
     dest_act = None
     dest_unsignup = False
+    send_emails = False
```
Comment (anonymoose2) [inline intranet/apps/eighth/views/admin/scheduling.py:397]:
```suggestion
```

**Review by anonymoose2 — CHANGES_REQUESTED** [review]

**Review by etnguyen03 — APPROVED** [review]
LGTM

**Review by anonymoose2 — APPROVED** [review]
LGTM
